key-manager/src/lib.rs

```diff
@@ -0,0 +1,355 @@
+    pub async fn load_secret(&mut self, epoch: u64) -> Result<(), Error> {
+        match self.secret_retriever.get(epoch).await? {
+            SecretState::Current(ikm) => self.insert_prk(ikm),
+            SecretState::Reconfiguration { old, new } => {
+                self.insert_prk(old);
+                self.insert_prk(new);
+            }
+        }
+        Ok(())
+    }
+
+    /// Derive an encryption key for the given [`sled_hardware::DiskIdentity`]
+    pub async fn disk_encryption_key(
+        &mut self,
+        epoch: u64,
+        disk_id: &DiskIdentity,
+    ) -> Result<VersionedAes256GcmDiskEncryptionKey, Error> {
+        let prk = if let Some(prk) = self.prks.get(&epoch) {
+            prk
+        } else {
+            self.load_secret(epoch).await?;
+            self.prks.get(&epoch).unwrap()
+        };
+
+        let mut key = Aes256GcmDiskEncryptionKey::default();
+
+        // Unwrap is safe because we know our buffer is large enough to hold the output key
+        prk.expand_multi_info(
+            &[
+                b"U.2-zfs-",
+                disk_id.vendor.as_bytes(),
+                disk_id.model.as_bytes(),
+                disk_id.serial.as_bytes(),
+            ],
+            key.0.as_mut(),
+        )
+        .unwrap();
+
+        Ok(VersionedAes256GcmDiskEncryptionKey { epoch, key: Secret::new(key) })
+    }
+
+    /// Return the epochs for all secrets which are loaded
+    pub fn loaded_epochs(&self) -> Vec<u64> {
+        self.prks.keys().copied().collect()
+    }
+
+    /// Clear the PRKs
+    ///
+    /// TODO(AJS): From what I can tell, the internal PRKs inside
+    /// Hkdf instances are not zeroized, and so will remain in memory
+    /// after drop. We should fix this one way or another.
+    pub fn clear(&mut self) {
+        self.prks = BTreeMap::new();
+    }
+
+    fn insert_prk(&mut self, ikm: VersionedIkm) {
+        let prk =
+            Hkdf::new(Some(&ikm.salt), ikm.ikm.expose_secret().0.as_ref());
+        self.prks.insert(ikm.epoch, prk);
+    }
+}
+
+/// The current state returned from a [`SecretRetriever`]
+pub enum SecretState {
+    /// A reconfiguration is not ongoing
+    Current(VersionedIkm),
+
+    /// A reconfiguration is ongoing
+    Reconfiguration { old: VersionedIkm, new: VersionedIkm },
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum SecretRetrieverError {
+    #[error("Secret does not exist for {0}")]
+    NoSuchEpoch(u64),
+}
+
+/// A mechanism for retrieving a secrets to use as input key material to HKDF-
+/// Extract.
+#[async_trait]
+pub trait SecretRetriever {
+    /// Return the latest secret
+    ////
+    /// This is useful when a new entity is being encrypted and there is no need
+    /// for a reconfiguration. When an entity is already encrypted, and needs to
+    /// be decrypted, the user should instead call the [`SecretRetriever::get`].
+    async fn get_latest(&self) -> Result<VersionedIkm, SecretRetrieverError>;
```

Comment:
Out of curiosity, why isn't this returning an epoch? (... which could potentially be passed to `Self::get`?)

What happens if this is called while a reconfiguration is happening?

Reply:
`VersionedIkm` contains the epoch.

What happens if this is called when reconfiguration is happening is the same as when it isn't. The latest secret will be returned as that's what the caller asked for. In general this should only be used for configuring a new device and so the latest is what the user actually wants. They have no use for older secrets.

In general, we can't deal with reconfiguration entirely within this part of the code. Even the call to `get` is only optimistic. There's an inherent race condition where a reconfiguration can be started after the caller gets the values of secrets back. It's the responsibility of the system supplying the input key material to ensure that that is fine and keys can be rotated one after another for any still available epochs (up to a policy limit).  In trust quorum, where we expect to use this, the sled-agent will get informed of a new configuration and trigger a rekey again.